src/testharness/TestHarness.cpp

```diff
@@ -119,14 +176,41 @@ bool TestHarness::runTestsForToolChain(std::string exeName, std::string tcName)
             << "\n";
 
   for (auto& test : invalidTests) {
-    std::cout << "  Skipped: " << test->getTestPath().filename().stem() << std::endl
-              << "  Error: " << Colors::YELLOW << test->getParseErrorMsg() << Colors::RESET << "\n";
+    std::cout << "  Skipped: " << test.first->getTestPath().filename().stem() << std::endl
+              << "  Error: " << Colors::YELLOW << test.first->getParseErrorMsg() << Colors::RESET << "\n";
   }
   std::cout << "\n";
 
+  std::cout << Colors::GREEN << "Completed Tests" << Colors::RESET << std::endl;
+  std::cout << "Hold on while we clean up any remaining threads, this might take a moment" << std::endl;
+
   return failed;
 }
 
+void TestHarness::threadRunTestsForToolChain(std::string tcName, std::string exeName, std::reference_wrapper<SubPackage> subPackage) {
+  std::cout << "Thread " << std::this_thread::get_id() << ": executing subpackage" << std::endl;
+  ToolChain toolChain = cfg.getToolChain(tcName); // Get the toolchain to use.
+  const fs::path& exe = cfg.getExecutablePath(exeName); // Set the toolchain's exe to be tested.
+  toolChain.setTestedExecutable(exe);
+
+  // set the runtime
+  if (cfg.hasRuntime(exeName)) // If we have a runtime, set that as well.
+    toolChain.setTestedRuntime(cfg.getRuntimePath(exeName));
+  else
+    toolChain.setTestedRuntime("");
+
+  for (size_t i = 0; i < subPackage.get().size(); ++i) {
+    std::unique_ptr<TestFile>& test = subPackage.get().at(i).first;
+    if (test->getParseError() == ParseError::NoError) {
+
+      TestResult result = runTest(test.get(), toolChain, cfg);
+      // keep the result with the test for pretty printing
+      std::optional<TestResult> res_clone = std::make_optional(result.clone());
```

Comment:
There's no need to clone here, since make_optional is just doing the same thing under the hood. `std::optional` doesn't store a reference (unless you're asking it to), so `std::make_optional` is going to call `TestResult(const TestResult &)` as part of initializing the stored value anyways.

In any case, if this is changed to promise/future, this line isn't needed anyways, since the next line should just change to using `std::promise::set_value`. No need to clone again.